### PR TITLE
avoid nan and inf in weight/label/init_score

### LINKED ISSUE #2359: Check weights for NANs
## Environment info

Operating System: AWS RHEL 7

CPU/GPU model: AWS ec2 m5.24xlarge

C++/Python/R version: Python 3.7

LightGBM version or commit hash: 2.2.3

### Set up variables
```python
# Set up datasets
x1 = pd.DataFrame(np.random.random(size=(100, 2))).values
y1 = np.log(x1[:, 0] ** 2 + x1[:, 1] ** 0.5) + np.random.normal(size=100) > 0
x2 = pd.DataFrame(np.random.random(size=(100, 2))).values
y2 = np.log(x2[:, 0] ** 2 + x2[:, 1] ** 0.5) + np.random.normal(size=100) > 0
w1 = np.random.random(size=100)
w2 = np.random.random(size=100)

param_dict = {
    'objective': 'binary',
     'tree_learner': 'feature',
     'metric': ['auc'],
     'learning_rate': 0.1,
     'max_bin': 127,
     'max_depth': 10,
     'num_leaves': 20,
     'min_data_in_leaf': 2,
     'num_threads': 48
}
```

### Training without weights
```python
# Training without weights
lgb_train = lgb.Dataset(x1, y1, free_raw_data=False)
lgb_val = lgb.Dataset(x2, y2, free_raw_data=False)

evals_result = {}
m_nw = lgb.train(
    params=param_dict, 
    train_set=lgb_train, 
    valid_sets=(lgb_val,), 
    num_boost_round=100, 
    early_stopping_rounds=10, 
    evals_result=evals_result
)
```
Returns
```
[1]	valid_0's auc: 0.584848
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's auc: 0.571717
[3]	valid_0's auc: 0.564242
[4]	valid_0's auc: 0.564646
[5]	valid_0's auc: 0.589495
[6]	valid_0's auc: 0.582424
[7]	valid_0's auc: 0.582828
[8]	valid_0's auc: 0.571515
[9]	valid_0's auc: 0.566465
[10]	valid_0's auc: 0.588283
Early stopping, best iteration is:
[5]	valid_0's auc: 0.589495
```

### Training with weights

```python
# Training with weights
lgb_train = lgb.Dataset(x1, y1, weight=w1, free_raw_data=False)
lgb_val = lgb.Dataset(x2, y2, weight=w2, free_raw_data=False)

evals_result = {}
m_w = lgb.train(
    params=params,
    train_set=lgb_train, 
    valid_sets=(lgb_val,), 
    num_boost_round=100, 
    early_stopping_rounds=5, 
    evals_result=evals_result
)
```
returns 
```
[1]	valid_0's auc: 0.5
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's auc: 0.5
[3]	valid_0's auc: 0.5
[4]	valid_0's auc: 0.5
[5]	valid_0's auc: 0.5
[6]	valid_0's auc: 0.5
Early stopping, best iteration is:
[1]	valid_0's auc: 0.5
```

In addition, the weight usage example here https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py#L18-L19 only used a constant weight of `1.0`, which is reset to `None` when the `lgb.Dataset` is initialized (https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L1356-L1357)

### PR DESCRIPTION
**Pull request by guolinke**
to fix #2359

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@guolinke I can try to add tests if you don't mind.

**Comment by StrikerRUS** [comment]
@guolinke During writing a test, I discovered that we need to update wrappers code (Python-package at least) as well.
```
import numpy as np
import lightgbm as lgb

from sklearn.datasets import load_boston

X, y = load_boston(True)
sequence = np.ones(y.shape[0])
sequence[0] = np.nan

lgb_data = lgb.Dataset(X, y).construct()
lgb_data.set_weight(sequence)
```
Now, `get_field()` and `get_*()` have inconsistent behavior: `get_field()` returns modified value, while `get_*()` - not.

```
lgb_data.get_weight()
>>> array([nan,  1.,  1.,  1.,  1., ...
```

```
lgb_data.get_field('weight')
>>> array([0.,  1.,  1.,  1.,  1., ...
```

**Comment by StrikerRUS** [comment]
I noticed that you used this PR to fix code style in `metadata.cpp`, so I replaced postfix increment with prefix one in b32988a370121f54ce736a3e6647102a598e95cb.

**Comment by guolinke** [comment]
@StrikerRUS 
Thanks, that is because .get_weight() will return the data in python, and .get_field will return data in cpp. This PR only fixes the invalid values in cpp side.

So maybe a better solution is to convert nan/inf in python side as well.

**Comment by StrikerRUS** [comment]
@guolinke OK! What do you think about the following plan? Let's merge this PR with cpp changes without tests then. And after that I'll propose Python side updates with tests.

**Comment by guolinke** [comment]
Sure! Sounds good
